torch/nn/parallel/scatter_gather.py

```diff
@@ -33,12 +33,13 @@ def scatter_map(obj):
 
 def scatter_kwargs(inputs, kwargs, target_gpus, dim=0):
     r"""Scatter with support for kwargs dictionary"""
-    inputs = scatter(inputs, target_gpus, dim) if inputs else []
-    kwargs = scatter(kwargs, target_gpus, dim) if kwargs else []
+    inputs = scatter(inputs, target_gpus, dim) if inputs else [()]
```

Comment:
Question, why this does not need to match `len(target_gpus)`? I mean should it be sth like:

```python
inputs = scatter(inputs, target_gpus, dim) if inputs else [() for _ in range(len(target_gpus))]
```

Reply:
If you make it match `len(target_gpus)` then you'll get multiple copies of the return value in the no-inputs case. That was what I meant by the SPMD support in the earlier comments. I'm not sure what the behavior should be here since in my own workflows I never use SPMD, and I gather it's going to be deprecated anyway.

For general cases, when you scatter the inputs and run the replicas, you get back the concatenated results from (D)DP which makes sense. When you provide no inputs though, I don't think you should get back the "concatenated" results because there were no inputs to scatter and it's just `n_replicas` copies of the stored parameter.

Would it be better to just raise an exception if you try to run a module with no inputs in SPMD mode (`len(target_gpus) > 1`)?

Reply:
@HapeMask Thanks for the explanation. 

> I never use SPMD, and I gather it's going to be deprecated anyway.

We plan to retire the SPMD mode in DDP. But `DataParallel` will stay intact, so it should still run in SPMD mode. 

IIUC and as you mentioned above, due to the following code, the current version of this PR would result in the replicate in `DataParallel` to return a single copy of the model, and hence it is not fully utilizing all GPUs using SPMD?

https://github.com/pytorch/pytorch/blob/4c99a9b6721c445a8c427020e54b63da0dbe622c/torch/nn/parallel/data_parallel.py#L154-L155

> Would it be better to just raise an exception if you try to run a module with no inputs in SPMD mode (len(target_gpus) > 1)?

I agree this is a weird mode, but I am not sure if we should raise in this case. Shall we add a warning in this case (both `inputs` and `kwargs` are None or empty) just to make sure this is indeed users intention? And I'd vote for still support SPMD for empty inputs and kwargs.

@ngimel @zhaojuanmao Please let us know if they think we should raise here. Thanks!